OdbcHook returns None. Related to #15016 issue.

Goodkat opened the pull request:
This PR is related to #15016 issue.
OdbcHook returns None for non-boolean-like string values in connect_kwargs dict arg, however connect_kwarg values should remain as is in this case.

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

Goodkat committed:
OdbcHook returns None. Related to #15016 issue.

This PR is related to #15016 issue.
OdbcHook returns None for non-boolean-like string values in connect_kwargs dict arg, however connect_kwarg values should remain as is in this case.

turbaszek commented on airflow/providers/odbc/hooks/odbc.py:
```diff
@@ -171,6 +171,8 @@ def clean_bool(val):  # pylint: disable=inconsistent-return-statements
                 elif val.lower() == 'false':
                     return False
+                else:
+                    return val
```
With this change we may be able to remove the pylint comment from L168, WDYT?

Goodkat replied:
There are different types (at least boolean and string) that could be returned by this function. Therefore (I suppose) the code analyser complains of this fact.
However, as I realized, it could be acceptable sometimes in dynamic languages, so we can conditionally return different data types.
In our case the clean_bool() function's result is used as a value in the dictionary and dictionaries itself can contain the keys/values of the different types. Therefore we may ignore this warning and remove the pylint comment from L168.
Please tell me if it does make sense for you.

Goodkat replied:
As the clean_bool is planned to be removed, the pylint comment will be cleaned as well.

Goodkat committed:
Removed clean_bool

According to the discussion on #15016, we agreed to remove the clean_bool function, as it is not needed actually for processing boolean values in JSON .

Goodkat committed:
Update odbc docs

Removed quotas for the boolean values accroding to the discussion on #15016 issue.

Goodkat committed:
Update odbc.py

Fixed the failing pydocstyle test: No blank lines allowed after function docstring (found 1)

Goodkat committed:
Update odbc.py

Fixed failing pylint check. R1721: Unnecessary use of a comprehension

Goodkat committed:
Update test_odbc.py

Integration Test has been changed as it should all work correctly with unquoted raw JSON booleans, which should make clean_bool unnecessary.

Goodkat committed:
Update test_odbc.py

Fixed input parameter {'ansi': true} -> {'ansi': True}

dstandish reviewed (commented):
I agree it's best remove this conversion.

I've [commented with reasoning in the issue](https://github.com/apache/airflow/issues/15016#issuecomment-837110088) and notes on how I think the logic came to be this way.

But I'm a little unsure about how we should release, given that it is a breaking change, howevery narrowly.  I think this requeires that we create a new major version of the odbc provider.

And it think we _could_ optionally first fix the `if` statement by adding an `else return val` in the current provider major version, before  removing `clean_bool` in next.  @turbaszek or @eladkal do you have advice here?  I'm not sure of the intricacies of bumping provider versions but would be happy to learn so that I can help out with this kind of thing.

-----

Some detail on the way in which this is "breaking":

From pyodbc.connect docstring:

```
      autocommit
        If False or zero, the default, transactions are created automatically as
        defined in the DB API 2.  If True or non-zero, the connection is put into
        ODBC autocommit mode and statements are committed automatically.
```

If someone is using `connect_kwargs` to set `autocommit` using the string value `'false'` then, before this change, autocommit would be `False`.  But after this change, it would be left as string `'False'`, and therefore it would be non-zero and non-False, and therefore the connection would be created with autocommit _on_.

On the other hand, users would **not** be affected if instead they managed autocommit by either (1) using DBApi methods, or (2) explicitly modifying the connection object, or (3) they used actual boolean values instead of strings.  And FWIW this hook has not been released very long.

In any case, I think **this PR needs a line in updating.md**, stating that this hook will no longer convert strings to boolean in `connect_kwargs`.

After that it's good to merge from my perspective, with the caveat that we need a little guidance from others on whether this needs to wait for next version.
